packages/keys/download-keys.ts

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env pnpm tsx
+/// <reference types="node" />
+
+import { spawn } from 'node:child_process';
+import fs from 'node:fs/promises';
+import path from 'node:path';
+import { createHash } from 'node:crypto';
+import { fileURLToPath } from 'node:url';
+import { version } from './package.json';
+import { Console } from 'node:console';
+
+const stdout = new Console(process.stdout);
+const stderr = new Console(process.stderr);
+const console = stderr;
+
+const cmd = (bin: string, args: string[], opt?: { cwd?: string }) =>
+  new Promise<void>((resolve, reject) => {
+    spawn(bin, args, { stdio: 'inherit', ...opt })
+      .on('spawn', () => console.log(bin, ...args))
+      .on('error', reject)
+      .on('exit', (code, signal) => {
+        if (code) process.exit(code);
+        if (signal) process.kill(process.pid, signal);
+        resolve();
+      });
+  });
+
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+let workspace: string | undefined;
+workspace =
+  (await import('workspace-root').then(
+    ({ workspaceRoot }) => workspaceRoot(),
+    () => null, // package not present, or no root
+  )) || undefined;
+
+const args = [undefined, ...process.argv];
+
+const outputOption = args[args.indexOf('--output') + 1];
+const versionOption = args[args.indexOf('--version') + 1];
+const clobberOption = args.includes('--clobber') || args.includes('--force');
+const rmFailedOption = args.includes('--rm-failed');
+const checksumsOption = args[args.indexOf('--checksums') + 1];
+const retryOption = args.includes('--retry');
+
+const versionTag = versionOption || 'v' + version;
+
+const ghRepoRoot = `https://github.com/penumbra-zone/penumbra/raw/${versionTag}`;
+const repoPath = 'crates/crypto/proof-params/src/gen';
+const remoteSource = path.join(ghRepoRoot, repoPath);
+
+const outDir = path.resolve(outputOption ?? path.join(workspace ?? __dirname, 'keys', versionTag));
+const checksumsFile = path.resolve(
+  checksumsOption ?? path.join(__dirname, `${versionTag}.sha256sum`),
+);
+
+const checksums = (await fs.readFile(checksumsFile))
+  .toString()
+  .split('\n')
+  .filter(id => id) // remove blank line
+  .map((line: string) => {
+    const [_line, sha256sum, name] = line.match(/^([0-9a-f]{64}) {2}([a-z_]+_pk.bin)$/)!;
+    return [sha256sum, name] as [string, string];
+  });
+
+const fileNames = checksums.map(([_, name]) => name);
+
+let retry = Number(retryOption);
+
+async function downloadKeys() {
+  // download
+  await cmd('curl', [
+    '--parallel',
+    '--create-dirs',
+    ...['--output-dir', outDir],
+    ...(clobberOption
+      ? // clobber will overwrite existing files
+        ['--clobber']
+      : // continue-at will resume or skip existing files
+        ['--continue-at', '-']),
+    '--location',
+    '--remote-name',
+    // curl supports url globbing
+    `${remoteSource}/{${fileNames.join(',')}}`,
+  ]);
+}
```

Comment:
just use curl instead of implementing all of these things

Reply:
nice approach!